lib/gc/Dialect/Microkernel/MicrokernelOps.cpp

```diff
@@ -8,7 +8,369 @@
+#undef FTAttr
+#undef ITAttr
+  if (!llvm::any_of(validDataTypes,
+                    [=](std::pair<TypeAttr, TypeAttr> type_pair) {
+                      return type_pair.first == dtypes[0] ||
+                             type_pair.second == dtypes[1];
+                    })) {
+    return op->emitOpError() << "invalid data type pair";
+  }
+
+  return success();
+}
+
+static LogicalResult verifyBrgemmFlags(ArrayAttr flags, BrgemmDispatchOp op,
+                                       const std::string_view &flagsName) {
+  // Verify flags.
+  if (failed(verifyUniquenessAndConsistency<BrgemmFlags>(flags, op, flagsName)))
+    return failure();
+
+  bool strideSet = false;
+  bool listSet = false;
+  for (auto flag : flags) {
+    if (cast<BrgemmFlagsAttr>(flag).getValue() == BrgemmFlags::STRIDE) {
+      strideSet = true;
+    }
+    if (cast<BrgemmFlagsAttr>(flag).getValue() == BrgemmFlags::LIST) {
+      listSet = true;
+    }
+  }
+  // VNNI flags must be specified only for bf16 type
+  if (strideSet && listSet) {
+    return op->emitOpError()
+           << "stride and addr flags conflict with each other";
+  }
+
+  return success();
+}
+
+LogicalResult BrgemmDispatchOp::verify() {
+  BrgemmDispatchOp &op = *this;
+  // 'inputs' = [m, n, k, lda, ldb, ldc, stride_a, stride_b] for BRGEMM.
+  size_t expected = 8;
+  size_t numInputs = op.getInputs().size();
+  if (numInputs != expected) {
+    return op.emitOpError()
+           << "expect " << expected << " args but got: " << numInputs;
+  }
+  // Verify data types
+  if (failed(verifyBrgemmDataTypes(op.getDataType(), op))) {
+    return failure();
+  }
+
+  // Verify leading dims.
+  ArrayRef<int64_t> inputs = op.getInputs();
+  int64_t n = inputs[1];
+  int64_t k = inputs[2];
+  int64_t lda = inputs[3];
+  int64_t ldb = inputs[4];
+  int64_t ldc = inputs[5];
+  if (lda < k)
+    return op.emitOpError() << "expect lda to be >= of dimension k\n";
+  if (ldb < n)
+    return op.emitOpError() << "expect ldb to be >= of dimension n\n";
+  if (ldc < n)
+    return op.emitOpError() << "expect ldc to be >= of dimension n\n";
+
+  // Verify dispatch flags.
+  return verifyBrgemmFlags(op.getFlags(), op, FLAGS_NAME);
+}
+
+/////////////////////////////////////////////////////
+// Start of BrgemmOp
+
+// TODO(haixin): could use compiler-wide VNNI utils?
+static bool isInVnniLayout(MemRefType memref) {
+  if (!memref.getElementType().isBF16() &&
+      !memref.getElementType().isSignedInteger(8) &&
+      !memref.getElementType().isUnsignedInteger(8)) {
+    return false;
+  }
+
+  auto blockingFactor = 0;
+  if (memref.getElementType().isBF16()) {
+    blockingFactor = 2;
+  } else if (memref.getElementType().isSignedInteger(8) ||
+             memref.getElementType().isUnsignedInteger(8)) {
+    blockingFactor = 4;
+  }
+  return memref.getShape().back() == blockingFactor;
+}
+
+static bool isTypeCompatible(Type outType, Type operandAType,
+                             Type operandBType) {
```

Comment:
This seems to check something more specific than type compatibility. Could you please reflect that with a name change?

Reply:
This function checks whether the input types of Memrefs are supported, currently the following types are supported:
1. f32xf32 -> f32
2. bf16xbf16 -> f32
3. [s8, u8]x[s8, u8] -> s32
Change the function name to `isTypeSupported`